File: src/pascal/heapsort.pas
program HeapSort240616;

const
  MAX = 10;

type
  Tgiven  = integer;
  Tvector = array[0..MAX] of Tgiven;

var
  vector : Tvector;

procedure auto(var vector:Tvector);
var
  i : integer;
begin
  for i := 0 to MAX-1 do
  begin
    vector[i]:=random(150);
  end;
end;

procedure display( vector: Tvector );
var
  i: integer;
begin
	writeln;
	write('=>	| ');
  for i := 0 to MAX-1 do write(vector[i],' | ');
  writeln;
end;

procedure swap( var a, b: integer );
var
  temp: integer;
begin
  temp := a;
  a := b;
  b := temp;
end;

procedure sink( var vector : Tvector; start, last: integer );
var
  root, child: integer;
begin
  root := start;
  while ( root * 2 + 1 <= last ) do begin
  child := root * 2 + 1;
  if ( child < last ) and ( vector[child] < vector[child + 1] ) then
    child := child + 1;
  if ( vector[root] < vector[child] ) then begin
    swap ( vector[root], vector[child] );
    root := child;
  end else
    break;
  end;
end;

procedure heighten( var vector : Tvector; count: integer );
var
  start: integer;
begin
  start := (count - 1) div 2;
  while ( start >= 0 ) do
  begin
    sink (vector, start, count-1);
    start := start - 1;
  end;
end;

procedure heapSort( var vector : Tvector; lastVector:integer );
var
  last : integer;
begin
  heighten(vector, MAX );
  last := MAX - 1;
  writeln;
  writeln('Heap');
  display(vector);
  while ( last > 0 ) do
  begin
    swap(vector[last], vector[0]);
    last := last - 1;
    sink(vector, 0, last);
  end;
end;

//program
begin
	writeln;
  auto(vector);
  writeln('Original Vector');
  display(vector);
  heapSort (vector, MAX);
  writeln;
  writeln('Sorted Vector');
  display(vector);
end.

File: src/c/Heapsort.c
#include <stdio.h>

void heapify(int arr[], int n, int i) {
  int largest = i;
  int left = 2 * i + 1;
  int right = 2 * i + 2;

  if (left < n && arr[left] > arr[largest])
    largest = left;

  // If the right child is larger than the largest so far
  if (right < n && arr[right] > arr[largest])
    largest = right;

  // If the largest element is not the root
  if (largest != i) {
    // Swap the largest element with the root
    int temp = arr[i];
    arr[i] = arr[largest];
    arr[largest] = temp;

    // Recursively heapify the affected sub-tree
    heapify(arr, n, largest);
  }
}

void heapsort(int arr[], int n) {
  // Build a max heap
  int i;
  for (i = n / 2 - 1; i >= 0; i--)
    heapify(arr, n, i);

  // One by one extract elements from the heap
  for (i = n - 1; i > 0; i--) {
    // Move the current root to the end
    int temp = arr[0];
    arr[0] = arr[i];
    arr[i] = temp;

    // Call max heapify on the reduced heap
    heapify(arr, i, 0);
  }
}

int main() {
  int arr[] = {12, 21, 13, 5, 1, 7};
  int i;
  int n = sizeof(arr) / sizeof(arr[0]);

  printf("Original array: ");
  for (i = 0; i < n; i++)
    printf("%d ", arr[i]);

  heapsort(arr, n);

  printf("\nSorted array: ");
  for (i = 0; i < n; i++)
    printf("%d ", arr[i]);

  return 0;
}
